File: application/usecases/src/main/java/org/acme/app/query/impl/GetCustomersQueryImpl.java
package org.acme.app.query.impl;

import org.acme.app.query.GetCustomersQuery;
import org.acme.domain.customer.CustomerGateway;

import javax.inject.Singleton;
import javax.transaction.Transactional;
import java.util.List;
import java.util.stream.Collectors;

@Singleton
@Transactional
public class GetCustomersQueryImpl implements GetCustomersQuery {
    private CustomerGateway customerGateway;

    public GetCustomersQueryImpl(CustomerGateway customerGateway) {
        this.customerGateway = customerGateway;
    }

    @Override
    public List<Response> getCustomers() {
        return customerGateway.getAllCustomers().stream()
                .map(customer -> new GetCustomersQuery.Response(customer.getName()))
                .collect(Collectors.toList());
    }
}

File: application/usecases/src/main/kotlin/org/acme/app/query/impl/GetCustomersQueryImpl.kt
package org.acme.app.query.impl

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.inject.Singleton
import jakarta.transaction.Transactional
import org.acme.app.query.GetCustomersQuery
import org.acme.domain.customer.CustomerGateway
import java.util.stream.Collectors

@ApplicationScoped
@Transactional
class GetCustomersQueryImpl @Inject constructor(private val customerGateway: CustomerGateway) : GetCustomersQuery {
    override fun customers(): List<GetCustomersQuery.Response> {
        return customerGateway.allCustomers()
            .map { GetCustomersQuery.Response(it.id, it.name) }
    }
}
